File: 백준/Gold/1062. 가르침/가르침.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static int N, K;
    static int[] words;
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());

        words = new int[N];
        for(int n = 0; n < N; ++n) {
            String word = br.readLine();
            words[n] = string2Bit(word);
        }

        int bit = 0;
        if(K > 0) bit |= (1 << 'a' - 'a'); 
        if(K > 1) bit |= (1 << 'n' - 'a'); 
        if(K > 2) bit |= (1 << 't' - 'a'); 
        if(K > 3) bit |= (1 << 'i' - 'a'); 
        if(K > 4) bit |= (1 << 'c' - 'a'); 

        System.out.println(combination(1, 5, bit));
    }

    static int combination(int idx, int count, int bit) {
        if(count >= K) {
            return countWord(bit);
        }
        if(idx == 26) {
            return 0;
        }

        int max = 0;
        if((bit & (1 << idx)) == 0) {
            max = Math.max(combination(idx + 1, count + 1, bit | (1 << idx)), max);
        }

        return Math.max(combination(idx + 1, count, bit), max);
    }

    static int countWord(int bit) {
        if(K < 5) return 0;

        int count = 0;
        for(int word : words) {
            if((bit & word) == word) ++count;   
        }
        return count;
    }

    static int string2Bit(String a) {
        int bit = 0;

        for(int i = 0; i < a.length(); ++i) {
            bit |= (1 << a.charAt(i) - 'a');
        }
        
        return bit;
    }
}

File: 백준/Gold/1062. 가르침/가르침.cc
#define MAX_N 51
#define MAX_LEN 16
#include <cstdio>

using namespace std;

int N, K;
int words[MAX_N];

int combination(const int& idx, const int& count, const int& bit);
int str2bit(const char* word);
int countWord(const int& bit);
int max(int a, int b) { return a > b ? a : b; }

int main() {
	scanf("%d %d", &N, &K);
	
	char word[MAX_LEN];
	for (int n = 0; n < N; ++n) {
		scanf("%s", &word);
		words[n] = str2bit(word);
	}
	
	printf("%d\n", combination(1, 5, str2bit("antic")));
}

int combination(const int& idx, const int& count, const int& bit) {
	if (count >= K) return countWord(bit);
	if (idx == 26) return 0;

	int maxi = 0;
	if (!(bit & 1 << idx)) maxi = max(maxi, combination(idx + 1, count + 1, bit | 1 << idx));
	return max(maxi, combination(idx + 1, count, bit));
}

int countWord(const int& bit) {
	if (K < 5) return 0;
	int count = 0;

	for (int n = 0; n < N; ++n) {
		if ((bit & words[n]) == words[n]) ++count;
	}
	return count;
}

int str2bit(const char* word) {
	int bit = 0;

	for (int i = 0; word[i] != '\0'; ++i) {
		bit |= 1 << word[i] - 'a';
	}

	return bit;
}